# pp_simScore(1)

## NAME

pp_simScore - print similarity and alignments between protein profile and sequence on the standard output

## SYNOPSIS

*pp_simScore* [OPTIONS] --fasta=_protein-sequence-file_ --prfl=_protein-profile-file_

## DESCRIPTION

Algorithm for calculating the similarity score and the optimal alignments between a protein profile and a protein sequence. 
The algorithm can optional take intron positions into account. Print to standard output.

## OPTIONS

### Mandatory options
    
*-f*, *--fasta*=_file_::
  Protein sequence file in FASTA format.

*-p*, *--prfl*= _file_::
  Protein profile file.

### Additional options

*-g*, *--gap_inter*=_float_::
  Gap costs for an alignment column that is a gap in an inter-block section. Default Value: -5.0

*-b*, *--gap_intra*=_float_::
  Gap costs for an alignment column that is a gap in a block. Default Value: -50.0

*-r*, *--gap_intron*=_float_::
  Gap costs for an gap in intron positions. Default Value: -5.0

*-e*, *--epsilon_intron*=_float_::
  Pseudocount parameter epsilon1, the pseudocount is added to a relative intron frequency v/w with (v+epsilon1)/(w+epsilon1+epsilon2). Default Value: 0.0000001

*-n*, *--epsilon_noIntron*=_float_::
  Pseudocount parameter epsilon2, the pseudocount is added to a relative intron frequency v/w with (v+epsilon1)/(w+epsilon1+epsilon2). Default Value: 0.1

*-i*, *--intron_weight_intra*=_float_::
  Value that is added to an intron score for a match of intron positions in a block. Default Value: 5.0

*-t*, *--intron_weight_inter*=_float_::
  Value that is added to an intron score for a match of intron positions in an inter-block. Default Value: 5.0

*-a*, *--alignment*=_number_::
  Number of optimal alignments that are computed. Default Value: 1

*-o*, *--out*=_format_::
 Denotes the output format, the following output options are implemented: :::
  *score*     :::: output is the similarity score
  *matrix*    :::: output are similarity matrix and similarity score
  *alignment* :::: output are the computed alignments as 
   * Alignment representation of P as symbols of 
     {AminoAcid, gap symbol or number of amino acids in inter-block}
   * Alignment representation of argmax of B as symbols of 
      {argmax AminoAcid for aligned block column, gap symbol or inter-block length}
   * Frequency of amino acid of P in aligned block column of B, if alignment type is a match 
  *matrix+alignment* :::: output are similarity matrix,
                          similarity score and the computed
                          alignments in the format described above
  *db*                :::: output are the computed alignment
                           as list of alignment frames,
                           an element of the list consists of:
     * starting position of the first amino acid of the protein sequence that is included in the alignment frame
     * block number in which the alignment frame is located
     * index of the first block column that is included in the alignment frame
     * length of the frame (number of alignment columns)
     * alignment type: 'm', 's'. 'p' or '-'
  *bp*               :::: output is a list of translations from the index of a block
                          to the number of the block in the .prfl file
  *consents*         :::: output is the average of the argmax
                          of the block columns for the complete profile
  *interblock*       :::: output is a list of all inter-block distance intervals

  ::: Default Value: *score*

*-h*, *--help*::
   Produce help message.

## EXAMPLE

  pp_simScore --fasta=EDW03868.1.fa --prfl=EOG09150290.prfl --out=alignment

## AUTHORS

AUGUSTUS was written by M. Stanke, O. Keller, S. KÃ¶nig, L. Gerischer, L. Romoth and L.Gabriel.
